velox/functions/prestosql/ArrayMinMaxBy.cpp

```diff
@@ -0,0 +1,179 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <folly/container/F14Set.h>
+
+#include <glog/logging.h>
+#include <velox/type/SimpleFunctionApi.h>
+#include "velox/expression/EvalCtx.h"
+#include "velox/expression/Expr.h"
+#include "velox/expression/VectorFunction.h"
+#include "velox/functions/lib/LambdaFunctionUtil.h"
+#include "velox/functions/lib/RowsTranslationUtil.h"
+
+namespace facebook::velox::functions {
+namespace {
+
+template <bool isMaxBy>
+class ArrayMinMaxByFunction : public exec::VectorFunction {
+ public:
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    VELOX_CHECK_EQ(args.size(), 2);
+    context.ensureWritable(rows, outputType, result);
+
+    exec::LocalDecodedVector arrayDecoder(context, *args[0], rows);
+    auto& decodedArray = *arrayDecoder.get();
+
+    // Flatten input array.
+    auto flatArray = flattenArray(rows, args[0], decodedArray);
+
+    std::vector<VectorPtr> lambdaArgs = {flatArray->elements()};
+    auto newNumElements = flatArray->elements()->size();
+
+    SelectivityVector validRowsInReusedResult =
+        toElementRows<ArrayVector>(newNumElements, rows, flatArray.get());
+
+    VectorPtr newElements;
+    auto elementToTopLevelRows = getElementToTopLevelRows(
+        newNumElements, rows, flatArray.get(), context.pool());
+
+    // Loop over lambda functions and apply these to elements of the base array.
+    // In most cases there will be only one function and the loop will run once.
+    auto it = args[1]->asUnchecked<FunctionVector>()->iterator(&rows);
+    while (auto entry = it.next()) {
+      auto elementRows = toElementRows<ArrayVector>(
+          newNumElements, *entry.rows, flatArray.get());
+      auto wrapCapture = toWrapCapture<ArrayVector>(
+          newNumElements, entry.callable, *entry.rows, flatArray);
+
+      entry.callable->apply(
+          elementRows,
+          &validRowsInReusedResult,
+          wrapCapture,
+          &context,
+          lambdaArgs,
+          elementToTopLevelRows,
+          &newElements);
+    }
+
+    // New elements after the lambda function is applied
+    exec::LocalDecodedVector decodedElements(
+        context, *newElements, validRowsInReusedResult);
+
+    BufferPtr resultIndices = allocateIndices(rows.size(), context.pool());
+    auto* rawResultIndices = resultIndices->asMutable<vector_size_t>();
+
+    BufferPtr resultNulls = allocateNulls(rows.size(), context.pool());
+    auto* mutableNulls = resultNulls->asMutable<uint64_t>();
+
+    auto indices = getMaxOrMinIndices(rows, *flatArray, *newElements, context);
+    rows.applyToSelected([&](auto row) {
+      auto offset = flatArray->offsetAt(row);
+      if (decodedElements->isNullAt(offset)) {
```

Comment:
should this be decodedArray? I'm not sure why you're checking if the first element of the array is null 

and if that's the case, could you instead call
localResult->addNulls(decodedArray->nulls(), rows)

decodedArray->nulls() returns all the nulls after merging the encodings, and addNulls sets the nulls in localResult anywhere where rows is set and decodedArray->nulls() is null

Reply:
@kevinwilfong Thanks so much for the review! For this one, I was trying to cover the case where the input element in the array isn't null, but after the lambda is applied, it returns null. For instance an array of maps and the lambda returns the value at some key and the key does not exist in any of the maps, in which case I want the function to return null as per the Java implementation. But setting localResult nulls from decodedArray->nulls() doesn't seem to work for this case, how can I work with this for such cases?

Reply:
Changed to auto offset = indices[row]; since the above was looking at the offset of the first element. Fixed by looking at each max/min index instead